### PR TITLE
[Add] HISA-Paged Block Sparse MQA Attention

### PR DESCRIPTION
**Pull request by handsomeRobotSK**
# Paged Block Sparse MQA Attention — Opt1 Pipeline 优化

## 变更文件

- `examples/HISA/paged_block_sparse_mqa_attn_expert.py`

## 优化概要

对 decode 场景下的 Paged Block Sparse MQA Attention kernel 进行 Expert 模式流水线优化，消除不必要的标量同步开销，减少流水线填充时间，并增加 MTE2∥V 重叠。

### C scope（Cube 管线）

| 改动项 | 说明 |
|--------|------|
| 4×K L1 专用缓冲区 | 原来 2×K L1 ping-pong（K[0]/K[1] 复用），改为 k_l1_0..3 四个独立槽位，消除 L1 缓冲区复用导致的 MTE1→MTE2 信号开销 |
| 4×L0C 专用缓冲区 | l0c_0..3 四个独立槽位，每个 MMA 结果独享一个 L0C，消除 FIX↔M 争用 |
| DMA 重排序 | Wave 0 优先发出 K[0]+Q DMA，再发 K[1..3]；MTE2 队列顺序变为 K0,Q,K1,K2,K3。流水线填充时间从 5 个 DMA 缩减到 2（仅 K0+Q 需在 MTE1 开始前完成，K1..K3 与 staging/MMA 重叠） |
| 移除 MTE1→MTE2 信号 | 专用 L1 缓冲区不复用，删除 staging 后的 MTE1→MTE2 释放信号及对应的 destroy wait，共减 ~10 条标量指令 |
| 移除 `if token < total_tokens` 守卫 | decode 场景下该条件始终为真，删除不必要的条件分支 |
| 4 ws slots + per-block cross-flags | workspace 从 2 个改为 4 个，C→V 跨核信号从单一 CROSS_FLAG_C2V 改为 FLAG_BLK0..BLK3 四个 per-block 即时信号，V scope 可按 block 就绪顺序独立处理 |

**C scope 波次结构（6 waves）：**

```
W0: DMA K[0]+Q → K[1..3]（无 MTE1→MTE2 等待，直接发出）
W1: wait(K0,Q data ready) → Stage Q+K[0] → MMA K[0]
W2: L0C0→ws[0]→BLK0 | Stage K[1] | MMA K[1]
W3: L0C1→ws[1]→BLK1 | Stage K[2] | MMA K[2]
W4: L0C2→ws[2]→BLK2 | Stage K[3] | MMA K[3]
W5: L0C3→ws[3]→BLK3（drain）
```

### V scope（Vector 管线）

| 改动项 | 说明 |
|--------|------|
| 双缓冲 s_ub_e / s_ub_l | 每个 AIV 独立持有 early/late 两份 s_ub，DMA 与计算可重叠 |
| Tail-fill 替代 mask pipeline | 用标量循环 `for i >= limit: logits[i] = -inf` 替代原来的向量化 mask pipeline（createvecindex→compare→bitwise_and→select），简化实现 |
| 移除 V→MTE2 信号 | 专用 UB 缓冲区不复用，删除 V→MTE2 方向的 buffer availability 信号（3 个 set_flag + 3 个 wait_flag = 6 条标量指令） |
| MTE2∥V 重叠 | 晚期 DMA 在早期 tail-fill+输出之前提交，MTE2 在 V 处理早期输出期间并行搬运晚期数据 |

**V scope 流程：**

```
early: DMA_in → relu → row_expand_mul → reduce_sum → tail-fill → enqueue late DMA → early output
late:  [wait DMA] → relu → row_expand_mul → reduce_sum → tail-fill → late output
              ↑ late MTE2 DMA 与 V 早期 tail-fill+output 重叠 ↑
```

### 测试改动

- 随机张量先在 CPU 上生成再 H2D 拷贝到 NPU，规避当前 CANN 版本的动态 kernel 初始化失败（aclnnInplaceRandom 等）
- `--batch` 默认值从 8 改为 1（decode 场景）

## 性能影响

| 指标 | 原始 | Opt1 |
|------|------|------|
| 流水线填充 | 5 DMA 必须全部完成 | 仅 K0+Q（2 DMA） |
| C scope 标量同步 | MTE1→MTE2 完整双向 | 仅 MTE2→MTE1（数据就绪方向） |
| V scope 标量同步 | V↔MTE2 完整双向 | 仅 MTE2→V（数据就绪方向） |
| C→V 跨核信号 | 1 个汇总信号 | 4 个 per-block 即时信号 |

## 不变量

- Grid 不变：batch × topk_groups
- 输出 shape 不变
- Reference 函数不变
- 精度验证：`torch.testing.assert_close(rtol=1e-2, atol=1e-2)`

### CODE REVIEW

**Review by LLMZhangYC — APPROVED** [review]
lgtm

**Review by benyang0506 — APPROVED** [review]
lgtm
